domain/keymanager/service/service.go

```diff
@@ -0,0 +1,256 @@
+
+	if len(keys) == 0 {
+		return nil
+	}
+
+	toAdd := make([]keymanager.PublicKey, 0, len(keys))
+	for i, keyToAdd := range keys {
+		parsedKey, err := ssh.ParsePublicKey(keyToAdd)
+		if err != nil {
+			return fmt.Errorf(
+				"%w %q at index %d: %w",
+				keyserrors.InvalidPublicKey, keyToAdd, i, err,
+			)
+		}
+
+		if reservedPublicKeyComments.Contains(parsedKey.Comment) {
+			return fmt.Errorf(
+				"public key %q at index %d contains a reserved comment %q that cannot be used: %w",
+				keyToAdd,
+				i,
+				parsedKey.Comment,
+				errors.Hide(keyserrors.ReservedCommentViolation),
+			)
+		}
+
+		toAdd = append(toAdd, keymanager.PublicKey{
+			Comment:     parsedKey.Comment,
+			Fingerprint: parsedKey.Fingerprint(),
+			Key:         keyToAdd,
+		})
+	}
+
+	return s.st.AddPublicKeysForUser(ctx, userID, toAdd)
+}
+
+// DeletePublicKeysForUser removes the keys associated with targets from the
+// users list of public keys. Targets can be an arbitrary list of a
+// public key fingerprint (sha256), comment or full key value to be
+// removed. Where a match is found the key will be removed. If no key exists for
+// a target this will result in no operation. The following errors can be
+// expected:
+// - [errors.NotValid] when the user id is not valid
+// - [accesserrors.UserNotFound] when the provided user does not exist.
+func (s *Service) DeleteKeysForUser(
+	ctx context.Context,
+	userID user.UUID,
+	targets ...string,
+) error {
+	if err := userID.Validate(); err != nil {
+		return fmt.Errorf(
+			"validating user id %q when deleting public keys: %w",
+			userID, err,
+		)
+	}
+
+	return s.st.DeletePublicKeysForUser(ctx, userID, targets)
+}
+
+// ImportPublicKeysForUser will import all of the public keys available for a
+// given subject and add them to the specified Juju user. If the user already
+// has one or more of the public keys being imported they will safely be skipped
+// with no errors being returned.
+// The following errors can be expected:
+// - [errors.NotValid] when the user id is not valid
+// - [github.com/juju/juju/domain/access/errors.UserNotFound] when the user does
+// not exist.
+// - [keyserrors.InvalidPublicKey] when a key being imported fails validation.
+// - [keyserrors.ReservedCommentViolation] when a key being added contains a
+// comment string that is reserved.
+// - [keyserrors.UnknownImportSource] when the source for the import operation
+// is unknown to the service.
+// - [keyserrors.ImportSubjectNotFound] when the source has indicated that the
+// subject for the import operation does not exist.
+func (s *Service) ImportPublicKeysForUser(
+	ctx context.Context,
+	userID user.UUID,
+	subject *url.URL,
+) error {
+	if err := userID.Validate(); err != nil {
+		return fmt.Errorf(
+			"validating user id %q when importing public keys from %q: %w",
+			userID, subject.String(), err,
+		)
+	}
+
+	keys, err := s.keyImporter.FetchPublicKeysForSubject(ctx, subject)
+	if errors.Is(err, importererrors.NoResolver) {
```

Comment:
I like switches for this kind of thing. No biggie.

Reply:
Done & Done